src/authn/http_basic.rs

```diff
@@ -0,0 +1,140 @@
+        // Load the file
+        let raw = tokio::fs::read_to_string(&authfile).await?;
+        // Parse the records into a map
+        let mut authmap = HashMap::new();
+        for line in raw.split_terminator('\n') {
+            let line = line.trim();
+            // Each line is username:{hash}value
+            let pair: Vec<&str> = line.splitn(2, ':').collect();
+            authmap.insert(pair[0].to_owned(), pair[1].to_owned());
+        }
+        // Attach the map to the struct
+        Ok(HttpBasic { authmap })
+    }
+
+    fn check_credentials(&self, username: String, password: String) -> bool {
+        // Note that it is consider a security risk to leak any information about
+        // why an auth failed. So returning a bool provides the minimal info necessary.
+        match self.authmap.get(&username) {
+            Some(ciphertext) => {
+                if ciphertext.starts_with("$2y$") {
+                    match bcrypt::verify(password, ciphertext) {
+                        Err(e) => {
+                            tracing::warn!(%e, "Error verifying bcrypted passwd");
+                            false
+                        }
+                        Ok(res) => res,
+                    }
+                } else {
+                    tracing::warn!("htpasswd has entries in the wrong format.");
+                    false
+                }
+            }
+            None => {
+                // Intentionally waste time to prevent timing attacks from disclosing
+                // the presence or absence of a user ID. The number of rounds ($07$) will
+                // control how long this takes. Higher is longer.
+                let _ = bcrypt::verify(
+                    username,
+                    "$2y$07$QCVM96JWmNWzx3k/7g1UXOLAO2y0imHGNjzEVkQoikrsV3gd4Xqk6",
+                );
+                false
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for HttpBasic {
+    // TODO: When Authz is plumbed in, we should be more specific.
+    type Item = Anonymous;
+
+    async fn authenticate(&self, auth_data: &str) -> anyhow::Result<Self::Item> {
+        if auth_data.is_empty() {
+            anyhow::bail!("Username and password are required")
+        }
+
+        let (username, password) = parse_basic(auth_data)?;
+        match self.check_credentials(username, password) {
+            true => Ok(Anonymous),
+            false => anyhow::bail!("Authentication failed"),
+        }
+    }
+}
+
+fn parse_basic(auth_data: &str) -> anyhow::Result<(String, String)> {
+    match auth_data.strip_prefix(HTTP_BASIC_PREFIX) {
+        None => anyhow::bail!("Wrong auth type. Only Basic auth is supported"),
+        Some(suffix) => {
+            // suffix should be base64 string
+            let decoded = String::from_utf8(base64::decode(suffix)?)?;
+            let pair: Vec<&str> = decoded.splitn(2, ':').collect();
+            Ok((pair[0].to_owned(), pair[1].to_owned()))
```

Comment:
We should validate length above here to avoid a DOS attack. If I send a base64 encoded string with no `:` separator, `pair[1]` will panic

Reply:
fixed